storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogLeaderEpochState.java

```diff
@@ -100,17 +100,29 @@ void handleSegmentWithCopySegmentStartedState(RemoteLogSegmentId remoteLogSegmen
 
     void handleSegmentWithCopySegmentFinishedState(Long startOffset, RemoteLogSegmentId remoteLogSegmentId,
                                                    Long leaderEpochEndOffset) {
+        // If there are duplicate segments uploaded due to leader-election, then mark them as unreferenced.
+        // Duplicate segments can be uploaded when the previous leader had tier-lags and the next leader uploads the
+        // segment for the same leader-epoch which is a super-set of previously uploaded segments.
+        // (eg)
+        // case-1: Duplicate segment
+        //      L0 uploaded segment S0 with offsets 0-100 and L1 uploaded segment S1 with offsets 0-200.
+        //      We will mark the segment S0 as duplicate and add it to unreferencedSegmentIds.
+        // case-2: Overlapping segments
+        //      L0 uploaded segment S0 with offsets 10-90 and L1 uploaded segment S1 with offsets 5-100, S2-101-200,
+        //      and so on. When the consumer request for segment with offset 95, it should get the segment S1 and not S0.
+        Map.Entry<Long, RemoteLogSegmentId> lastEntry = offsetToId.lastEntry();
+        while (lastEntry != null && lastEntry.getKey() >= startOffset && highestLogOffset <= leaderEpochEndOffset) {
```

Comment:
Thinking more on this, we may have to update the logic to find the [copiedOffset](https://github.com/apache/kafka/blob/trunk/core/src/main/java/kafka/log/remote/RemoteLogManager.java#L499) such  that it's minimum of (end_offset_for_that_epoch_in_checkpoint_file, highest_remote_offset). In this case, when B2 was leader, copied_offset should be min(80, 112) where 80 is end_offset for LE0 and 112 is the highest_remote_offset for LE0.

So, that B2 can able to upload the segment S2 (12:101)

Reply:
Hi, thank you for explanation, sounds reasonable.

Sorry for being too questionable. One more point:
What could happen if we instead:
```
while (lastEntry != null && lastEntry.getKey() >= startOffset && highestLogOffset <= leaderEpochEndOffset) {
```
will do
```
while (lastEntry != null && lastEntry.getKey() >= startOffset)
```

How I see it:
- In case we sent twice same segment [somehow], It will be replaced with the same one
- In case unclean.leader.election set to true, we are going to have up to date data, in that case TS will work same as local storage function wise [data loss due to dead leader]. Less discrepancy - easier to maintain and understand
- I do believe we do not send from leader and all the followers same segments. The leader is someone who sends data to tiered storage, so we expect at the same time send only 1 segment for partition-X, in case second  request comes for same partition, it means that new leader is elected and it is in charge and its data is more valuable.
- It is possible that even ISR won't be synced fully `replica.lag.time.max.ms` [default= 30s] allows that, it means that it is possible that few messages wont be synced and potentially deleted in case traffic stops coming few messages after Leader broker died and new leader got them. What if Broker 1 uploaded segment [0, 100], Broker-2 was insync by replicated [0,90]. Broker-1 died, Broker-2 elected as a Leader and got 5 messages, Tiered storage won't be in charge of those 5 messages and segment could be deleted. After traffic being restored new segment is going to roll and its offset starts with 96, when last offset in TS is 100, so we delete segment and push new one starting from 96, and we could have a gap afterwards. 

Of course this example is an edge case and configuration should be quite agressive for that in terms for retention times etc. But are the any drawbacks by removing `highestLogOffset <= leaderEpochEndOffset` from the while? What kind of problem we can face and why having this check is more safe rather than remove it?

Reply:
Test example
```
@Test
    void handleSegmentWithCopySegmentFinishedStateForDuplicateSegmentsWithSecondSegmentEarlierEndOffset() {
        // Broker-1 is a Leader for Partition-0, Broker-2 is follower
        RemoteLogSegmentId segmentId1BRoker1 = new RemoteLogSegmentId(tpId, Uuid.randomUuid());
        RemoteLogSegmentId segmentId1Broker2 = new RemoteLogSegmentId(tpId, Uuid.randomUuid());

        //Leader and follower are in sync
        epochState.handleSegmentWithCopySegmentFinishedState(10L, segmentId1BRoker1, 100L);
        epochState.handleSegmentWithCopySegmentFinishedState(10L, segmentId1Broker2, 100L);


        //Broker-1 received some messages and part of them were replicated by follower
        // Broker-2 was not able to replicate all the messages because Leader is died. But were able to upload segment before death [101, 200]
        RemoteLogSegmentId segmentId2Broker1 = new RemoteLogSegmentId(tpId, Uuid.randomUuid());
        epochState.handleSegmentWithCopySegmentFinishedState(101L, segmentId2Broker1, 200L);
        
        // Broker-2 still in-sync according to `replica.lag.time.max.ms`
        // Last offset in sync is 150L
        // Broker-2 gets the leadership and fills the segment 101L-190L and uploads to Tiered storage
        RemoteLogSegmentId segmentId2Broker2 = new RemoteLogSegmentId(tpId, Uuid.randomUuid());
        epochState.handleSegmentWithCopySegmentFinishedState(101L, segmentId2Broker2, 190L);
        
        //Traffic stops
        
        //Since LeaderEpoch was changed, I expect that more fresh data will be on the TS
        assertEquals(190L, epochState.highestLogOffset());
        //segment 2 uploaded by Broker-1 is expected to be unreferenced
        assertTrue(epochState.unreferencedSegmentIds().contains(segmentId2Broker1));
        assertEquals(1, epochState.unreferencedSegmentIds().size());
    }
    ```

Reply:
Why would segment data for [151L,190L] is lost? This will be part of the next leader epoch state(viz 1) which should contain this segment. @kamalcph @Nickstery  Am I missing anything here?

Reply:
Thanks @kamalcph for the detailed explanation. It is better to return the epoch along with highest-offset and then check the minimum of the respective end offset in the local leader epoch chain for that epoch and the highest offset uploaded to remote storage for that epoch.

Reply:
Filed https://issues.apache.org/jira/browse/KAFKA-15241 to followup on this.